packages/cw721/src/traits.rs

```diff
@@ -1,168 +1,64 @@
-        deps: DepsMut,
-        env: Env,
-        info: MessageInfo,
-        spender: String,
-        token_id: String,
-    ) -> Result<Response<C>, Self::Err>;
-
-    fn approve_all(
-        &self,
-        deps: DepsMut,
-        env: Env,
-        info: MessageInfo,
-        operator: String,
-        expires: Option<Expiration>,
-    ) -> Result<Response<C>, Self::Err>;
-
-    fn revoke_all(
-        &self,
-        deps: DepsMut,
-        env: Env,
-        info: MessageInfo,
-        operator: String,
-    ) -> Result<Response<C>, Self::Err>;
-
-    fn burn(
-        &self,
-        deps: DepsMut,
-        env: Env,
-        info: MessageInfo,
-        token_id: String,
-    ) -> Result<Response<C>, Self::Err>;
-}
-
-pub trait Cw721Query<T>
-where
-    T: Serialize + DeserializeOwned + Clone,
-{
-    // TODO: use custom error?
-    // How to handle the two derived error types?
-
-    fn contract_info(&self, deps: Deps) -> StdResult<ContractInfoResponse>;
-
-    fn num_tokens(&self, deps: Deps) -> StdResult<NumTokensResponse>;
-
-    fn nft_info(&self, deps: Deps, token_id: String) -> StdResult<NftInfoResponse<T>>;
-
-    fn owner_of(
-        &self,
-        deps: Deps,
-        env: Env,
-        token_id: String,
-        include_expired: bool,
-    ) -> StdResult<OwnerOfResponse>;
-
-    fn operator(
-        &self,
-        deps: Deps,
-        env: Env,
-        owner: String,
-        operator: String,
-        include_expired: bool,
-    ) -> StdResult<OperatorResponse>;
-
-    fn operators(
-        &self,
-        deps: Deps,
-        env: Env,
-        owner: String,
-        include_expired: bool,
-        start_after: Option<String>,
-        limit: Option<u32>,
-    ) -> StdResult<OperatorsResponse>;
-
-    fn approval(
-        &self,
-        deps: Deps,
-        env: Env,
-        token_id: String,
-        spender: String,
-        include_expired: bool,
-    ) -> StdResult<ApprovalResponse>;
-
-    fn approvals(
+use std::fmt::Debug;
+
+use cosmwasm_std::{Deps, Empty, Env, MessageInfo};
+use schemars::JsonSchema;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::error::Cw721ContractError;
+
+/// This is an exact copy of `CustomMsg`, since implementing a trait for a type from another crate is not possible.
+///
+/// Possible:
+/// `impl<T> Cw721CustomMsg for Option<T> where T: Cw721CustomMsg {}`
+///
+/// Not possible:
+/// `impl<T> CustomMsg for Option<T> where T: CustomMsg {}`
+///
+/// This will be removed once the `CustomMsg` trait is moved to the `cosmwasm_std` crate: https://github.com/CosmWasm/cosmwasm/issues/2056
+pub trait Cw721CustomMsg: Serialize + Clone + Debug + PartialEq + JsonSchema {}
+
+pub trait Cw721State: Serialize + DeserializeOwned + Clone + Debug {}
+
+impl Cw721State for Empty {}
+impl<T> Cw721State for Option<T> where T: Cw721State {}
+impl Cw721CustomMsg for Empty {}
+impl<T> Cw721CustomMsg for Option<T> where T: Cw721CustomMsg {}
+
+pub trait StateFactory<S> {
+    fn create(
```

Comment:
Introducing new `StateFactory` with `create` and `validate` functions, for atomic state management. Avoids code duplication like during instantiation, execution, and migration.

Like here: ...